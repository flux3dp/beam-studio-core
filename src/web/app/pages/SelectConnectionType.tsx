import classNames from 'classnames';
import React, { useEffect, useMemo } from 'react';

import dialog from 'app/actions/dialog-caller';
import i18n from 'helpers/i18n';
import storage from 'implementations/storage';
import windowLocationReload from 'app/actions/windowLocation';

import styles from './SelectConnectionType.module.scss';

let lang = i18n.lang.initialize;
const updateLang = () => {
  lang = i18n.lang.initialize;
};

const TYPE_URL_MAP = {
  wifi: '#initialize/connect/connect-wi-fi',
  wired: '#initialize/connect/connect-wired',
  ether2ether: '#initialize/connect/connect-ethernet',
  usb: '#initialize/connect/connect-usb',
};

const SelectConnectionType = (): JSX.Element => {
  useEffect(() => {
    updateLang();
  }, []);

  const isNewUser = useMemo(() => !storage.get('printer-is-ready'), []);

  const handleBtnClick = () => {
    if (isNewUser) {
      storage.set('new-user', true);
    }
    storage.set('printer-is-ready', true);
    dialog.showLoadingWindow();
    window.location.hash = '#studio/beambox';
    windowLocationReload();
  };

  const handleConnectionTypeClick = (type: 'wifi' | 'wired' | 'ether2ether' | 'usb') => {
    const url = TYPE_URL_MAP[type];
    window.location.hash = url;
  };

  const renderConnectionTypeButton = (type: 'wifi' | 'wired' | 'ether2ether' | 'usb'): JSX.Element => (
    <button
      id={`connect-${type}`}
      type="button"
      className={classNames('btn', styles.btn)}
      onClick={() => handleConnectionTypeClick(type)}
    >
      {lang.connection_types[type]}
      {type === 'usb' ? <span className={styles.sub}>{lang.connect_usb.title_sub}</span> : null}
    </button>
  );

  const renderConnectionTypeContainer = (type: 'wifi' | 'wired' | 'ether2ether' | 'usb'): JSX.Element => (
    <div className={styles['btn-container']}>
      <img className={styles.icon} src={`img/init-panel/icon-${type}.svg`} draggable="false" />
      {renderConnectionTypeButton(type)}
    </div>
  );

  return (
    <div className={styles.container}>
      <div className={styles['top-bar']} />
      <div className={styles.btns}>
        <div className={classNames(styles.btn, styles.primary)} onClick={handleBtnClick}>
          {isNewUser ? lang.skip : lang.cancel}
        </div>
      </div>
      <div className={styles.main}>
        <h1 className={styles.title}>{lang.select_connection_type}</h1>
        <div className={styles.btns}>
          {renderConnectionTypeContainer('wifi')}
          {renderConnectionTypeContainer('wired')}
          {renderConnectionTypeContainer('ether2ether')}
          {renderConnectionTypeContainer('usb')}
        </div>
      </div>
    </div>
  );
};

export default SelectConnectionType;
