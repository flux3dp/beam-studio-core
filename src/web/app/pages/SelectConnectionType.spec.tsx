import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const mockShowLoadingWindow = jest.fn();
jest.mock('app/actions/dialog-caller', () => ({
  showLoadingWindow: mockShowLoadingWindow,
}));

const mockGet = jest.fn();
const mockSet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: mockGet,
  set: mockSet,
}));

jest.mock('helpers/i18n', () => ({
  lang: {
    initialize: {
      select_connection_type: 'How do you wish to connect?',
      connection_types: {
        wifi: 'Wi-Fi',
        wired: 'Wired Network',
        ether2ether: 'Direct Connection',
        usb: 'USB Connection',
      },
      connect_usb: {
        title_sub: ' (HEXA Only)',
      },
      skip: 'Skip',
      cancel: 'Cancel',
    },
  },
}));

const mockWindowLocationReload = jest.fn();
jest.mock('app/actions/windowLocation', () => mockWindowLocationReload);

// eslint-disable-next-line import/first
import SelectConnectionType from './SelectConnectionType';

describe('test Select-Connection-Type', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('should render correctly with new user', () => {
    mockGet.mockReturnValue(false);

    const wrapper = mount(<SelectConnectionType />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(mockGet).toHaveBeenCalledTimes(1);
    expect(mockGet).toHaveBeenNthCalledWith(1, 'printer-is-ready');
    expect(mockSet).not.toHaveBeenCalled();

    wrapper.find('.btn-action').at(0).simulate('click');
    expect(window.location.hash).toBe('#initialize/connect/connect-wi-fi');

    wrapper.find('.btn-action').at(1).simulate('click');
    expect(window.location.hash).toBe('#initialize/connect/connect-wired');

    wrapper.find('.btn-action').at(2).simulate('click');
    expect(window.location.hash).toBe('#initialize/connect/connect-ethernet');

    wrapper.find('.btn-action').at(3).simulate('click');
    expect(window.location.hash).toBe('#initialize/connect/connect-usb');

    wrapper.find('.btn-page').simulate('click');
    expect(mockSet).toHaveBeenCalledTimes(2);
    expect(mockSet).toHaveBeenNthCalledWith(1, 'new-user', true);
    expect(mockSet).toHaveBeenNthCalledWith(2, 'printer-is-ready', true);
    expect(mockShowLoadingWindow).toHaveBeenCalledTimes(1);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(window.location.hash).toBe('#studio/beambox');
    expect(mockWindowLocationReload).toHaveBeenCalled();
  });

  test('should render correctly but not new user', () => {
    mockGet.mockReturnValue(true);

    const wrapper = mount(<SelectConnectionType />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(mockGet).toHaveBeenCalledTimes(1);
    expect(mockGet).toHaveBeenNthCalledWith(1, 'printer-is-ready');
    expect(mockSet).not.toHaveBeenCalled();

    wrapper.find('.btn-page').simulate('click');
    expect(mockSet).toHaveBeenCalledTimes(1);
    expect(mockSet).toHaveBeenNthCalledWith(1, 'printer-is-ready', true);
  });
});
