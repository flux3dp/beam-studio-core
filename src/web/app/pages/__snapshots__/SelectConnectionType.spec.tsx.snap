// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SelectConnectionType should render correctly but not new user 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="top-bar"
    />
    <div
      class="btns"
    >
      <div
        class="btn primary"
      >
        Cancel
      </div>
    </div>
    <div
      class="main"
    >
      <h1
        class="title"
      >
        How do you wish to connect?
      </h1>
      <div
        class="btns"
      >
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wifi.svg"
          />
          <button
            class="btn btn"
            id="connect-wifi"
            type="button"
          >
            Wi-Fi
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wired.svg"
          />
          <button
            class="btn btn"
            id="connect-wired"
            type="button"
          >
            Wired Network
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-ether2ether.svg"
          />
          <button
            class="btn btn"
            id="connect-ether2ether"
            type="button"
          >
            Direct Connection
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-usb.svg"
          />
          <button
            class="btn btn"
            id="connect-usb"
            type="button"
          >
            USB Connection
             
            <span
              class="sub"
            >
               (HEXA Only)
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test SelectConnectionType should render correctly with new user 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="top-bar"
    />
    <div
      class="btns"
    >
      <div
        class="btn primary"
      >
        Skip
      </div>
    </div>
    <div
      class="main"
    >
      <h1
        class="title"
      >
        How do you wish to connect?
      </h1>
      <div
        class="btns"
      >
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wifi.svg"
          />
          <button
            class="btn btn"
            id="connect-wifi"
            type="button"
          >
            Wi-Fi
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wired.svg"
          />
          <button
            class="btn btn"
            id="connect-wired"
            type="button"
          >
            Wired Network
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-ether2ether.svg"
          />
          <button
            class="btn btn"
            id="connect-ether2ether"
            type="button"
          >
            Direct Connection
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-usb.svg"
          />
          <button
            class="btn btn"
            id="connect-usb"
            type="button"
          >
            USB Connection
             
            <span
              class="sub"
            >
               (HEXA Only)
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test SelectConnectionType should render correctly with new user 2`] = `
<div>
  <div
    class="container"
  >
    <div
      class="top-bar"
    />
    <div
      class="btns"
    >
      <div
        class="btn primary"
      >
        Skip
      </div>
    </div>
    <div
      class="main"
    >
      <h1
        class="title"
      >
        How do you wish to connect?
      </h1>
      <div
        class="btns"
      >
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wifi.svg"
          />
          <button
            class="btn btn"
            id="connect-wifi"
            type="button"
          >
            Wi-Fi
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wired.svg"
          />
          <button
            class="btn btn"
            id="connect-wired"
            type="button"
          >
            Wired Network
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-ether2ether.svg"
          />
          <button
            class="btn btn"
            id="connect-ether2ether"
            type="button"
          >
            Direct Connection
             
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-usb.svg"
          />
          <button
            class="btn btn"
            id="connect-usb"
            type="button"
          >
            USB Connection
             
            <span
              class="sub"
            >
               (HEXA Only)
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
